### PR TITLE
Ignore failures from unbottled formulae

### PR DESCRIPTION
**Pull request by carlocab**
This completes #687. It's not enough to call `skipped` because some
steps will still have a `:failed` status, and that will cause `test-bot`
to exit with an error.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
> steps will still have a `:failed` status, and that will cause `test-bot`
> to exit with an error.

Which ones? I'd rather those steps were also skipped rather than us ignoring failures.

**Comment by carlocab** [comment]
The ones we `skipped`. Doing `skipped` doesn't make the `:failed` status go away; it just adds the formula to the `@skipped_or_failed_formulae` array.

**Comment by MikeMcQuaid** [comment]
> Doing `skipped` doesn't make the `:failed` status go away

Yeh, I don't think it should. I think it should still show up as "FAILED". If the issue is the exit code: can we handle that instead?

**Comment by carlocab** [comment]
Hmmm, we could. Let me have a look.

**Comment by carlocab** [comment]
I haven't been able to work out how to handle the exit code without tweaking the `@status` a bit, but the changes I'm about to push should still show `FAILED` in log when something fails, because of

https://github.com/Homebrew/homebrew-test-bot/blob/b1e22363a0ea7a0abb793257c8762190c3bbea5d/lib/step.rb#L96-L97

and

https://github.com/Homebrew/homebrew-test-bot/blob/b1e22363a0ea7a0abb793257c8762190c3bbea5d/lib/step.rb#L61-L63

The next time `@status` is read is when determining the exit code, so it seems that that's the only sensible way to handle that.

**Comment by carlocab** [comment]
I believe the exit code is determined here:

https://github.com/Homebrew/homebrew-test-bot/blob/b1e22363a0ea7a0abb793257c8762190c3bbea5d/lib/test_bot.rb#L126

which comes from

https://github.com/Homebrew/homebrew-test-bot/blob/b1e22363a0ea7a0abb793257c8762190c3bbea5d/lib/test_runner.rb#L76

so I don't currently see a sensible way of changing the exit code without changing some steps to not become `:failed`. I've tried to minimise the steps I'm doing this for to the minimum required. I'll also add some output at the end to indicate that some failures were ignored.

**Comment by carlocab** [comment]
I've updated this to handle build/linkage/test failures for unbottled dependents too.

The idea is to downgrade all these errors into warnings that will still display failure messages but won't cause `test-bot` to exit with an error. I'm thinking it might be helpful to get @BrewTestBot to post comments on a PR about warnings in the CI log to indicate that they should be looked at, since no one will read the logs otherwise when CI is green.

**Comment by SMillerDev** [comment]
> I'm thinking it might be helpful to get @BrewTestBot to post comments on a PR about warnings in the CI log to indicate that they should be looked at, since no one will read the logs otherwise when CI is green.

I believe Jenkins used to do this and it made the issue overview so much easier.

**Comment by carlocab** [comment]
I was wondering why @BrewTestBot stopped doing that. We could maybe roll our own, but there's the issue of access to secrets on PR branches.

**Comment by Bo98** [comment]
> warnings in the CI log to indicate that they should be looked at, since no one will read the logs otherwise when CI is green.

Would GitHub Actions annotations help? Like the CI coverage annotations we have in Homebrew/brew.

**Comment by carlocab** [comment]
They would help, yes. Easier to miss than comments, but better than hidden away behind a green CI log.

**Comment by MikeMcQuaid** [comment]
Yeh, I think something like that would be ideal for adding warnings 👍🏻

File: lib/test_runner.rb [hunk 1]
```diff
@@ -53,12 +53,25 @@ def run!(tap, git:, args:)
...
+                           .compact
+      steps_output = if failed_steps.empty? && ignored_steps.empty?
         "All steps passed!"
       else
-        failed_steps_output = ["Error: #{failed_steps.count} failed #{"step".pluralize(failed_steps.count)}!"]
-        failed_steps_output += failed_steps.map(&:command_trimmed)
-        failed_steps_output.join("\n")
+        output_lines = []
+
+        unless ignored_steps.empty?
```
Comment (MikeMcQuaid) [inline lib/test_runner.rb:64]:
```suggestion
        if ignored_steps.present?
```

File: lib/test_runner.rb [hunk 2]
```diff
@@ -53,12 +53,25 @@ def run!(tap, git:, args:)
...
+          output_lines += ["Warning: #{ignored_steps.count} failed #{"step".pluralize(ignored_steps.count)} ignored!"]
+          output_lines += ignored_steps.map(&:command_trimmed)
+        end
+
+        unless failed_steps.empty?
```
Comment (MikeMcQuaid) [inline lib/test_runner.rb:69]:
```suggestion
        if failed_steps.present?
```

File: lib/test_runner.rb [hunk 3]
```diff
@@ -53,12 +53,25 @@ def run!(tap, git:, args:)
       failed_steps = tests.map(&:failed_steps)
                           .flatten
                           .compact
-      steps_output = if failed_steps.empty?
+      ignored_steps = tests.map(&:ignored_steps)
+                           .flatten
+                           .compact
+      steps_output = if failed_steps.empty? && ignored_steps.empty?
```
Comment (MikeMcQuaid) [inline lib/test_runner.rb:59]:
```suggestion
      steps_output = if failed_steps.blank? && ignored_steps.blank?
```

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**